### PR TITLE
Add longitudinal data

### PR DESCRIPTION
**Pull request by kohpangwei**

### CODE REVIEW

File: general_autoencoder.py [hunk 1]
```diff
@@ -24,18 +24,21 @@ def __init__(self,
...
         assert age_preprocessing_method in ['subtract_a_constant', 'divide_by_a_constant']
         self.age_preprocessing_method = age_preprocessing_method
-        self.include_age_in_encoder_input = include_age_in_encoder_input         
+        self.include_age_in_encoder_input = include_age_in_encoder_input   
         # include_age_in_encoder_input is whether age is used to approximate the posterior over Z. 
         # Eg, we need this for rate-of-aging autoencoder. 
         
         self.can_calculate_Z_mu = True # does the variable Z_mu make any sense for the model. 
         
+        self.uses_longitudinal_data = False # does the model accomodate longitudinal data as well.
```
[the remark this answers is not in the document]
Comment (tensorflownoob99) [reply general_autoencoder.py]: by analogy with the lines above but i agree it looks weird. if we're going to do this differently, should we change all the lines? (eg, self.can_calculate_Z_mu = can_calculate_Z_mu, and make can_calculate_Z_mu an input parameter).

[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: (sorry, that comment is actually from emma)

File: general_autoencoder.py [hunk 2]
```diff
@@ -24,18 +24,21 @@ def __init__(self,
...
-        include_age_in_encoder_input=False,     
+        include_age_in_encoder_input=False,  
+        uses_longitudinal_data=False,
         regularization_weighting_schedule={'schedule_type':'constant', 'constant':1}):
 
         self.need_ages = False # whether ages are needed to compute loss or other quantities. 
         assert age_preprocessing_method in ['subtract_a_constant', 'divide_by_a_constant']
         self.age_preprocessing_method = age_preprocessing_method
-        self.include_age_in_encoder_input = include_age_in_encoder_input         
+        self.include_age_in_encoder_input = include_age_in_encoder_input
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: this is already commented in the line below.

File: general_autoencoder.py [hunk 3]
```diff
@@ -230,7 +281,14 @@ def _fit_from_processed_data(self, train_data, valid_data, train_ages=None, vali
         
         print("Train size %i; valid size %i" % (
             self.train_data.shape[0], self.valid_data.shape[0]))
-                
+        
+        if self.uses_longitudinal_data:
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: we store all the other data a couple lines above; this seemed consistent.

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    on the longitudinal data, it calls the longitudinal loss function.
+    A hyperparameter, longitudinal_loss_weighting_factor, controls the relative weighting of the two losses. 
+    
+    Potential todo: right now this only computes the validation loss on cross-sectional data. 
+    """    
+    
+    def __init__(self,
+                 k_age,
+                 sparsity_weighting=0,
+                 aging_rate_scaling_factor=.1,
```
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: whooops yeah that's no good, good catch. i will check the other classes to make sure this isn't a problem anywhere else.

File: general_autoencoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kohpangwei) [reply general_autoencoder.py]: Yup, sure.

File: general_autoencoder.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kohpangwei) [reply general_autoencoder.py]: Yeah, I saw that. I still didn't understand how they're different, or when you might use one but not the other.

File: general_autoencoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kohpangwei) [reply general_autoencoder.py]: Ok, I guess the lines above might not be in the right place then. Probably not a big deal (and this might well be my fault in the original code) but I don't understand when we pass data around and when we keep it in the object. Seems good to either always pass data around, or put it in the object at the top level?

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                
+    def fill_feed_dict(self, 
+                       data, 
+                       regularization_weighting,
+                       ages, 
+                       idxs,
+                       age_adjusted_data=None):
+        
+        # The standard feed dict method for cross-sectional data. 
+        # We probably could just use the superclass method here?
```
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: i will leave as is for now because it is small.

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        return feed_dict
+    
+    def get_longitudinal_loss(self):
+        # loss function for longitudinal data. 
+        
+        # compute initial position in the latent space. 
+        Z0 = self.encode(self.X, self.ages)
+            
+        # now project Z0 forward to get Z1. This requires multiplying the age components by followup_ages / ages
+        Z1 = tf.concat([Z0[:, :self.k_age] * tf.reshape((1.0*self.followup_ages / self.ages), [-1, 1]), # broadcasting
```
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: changed this. as far as I know, Z should consistently refer to C * A. If we find places that is not the case, we should fix it?

File: general_autoencoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: yeah perhaps so. this feels peripheral to this PR so i will put it as an issue but leave as is for now.

File: variational_longitudinal_rate_of_aging_autoencoder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: yeah i agree. will delete.

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        Xr0 = self.decode(Z0)
+        Xr1 = self.decode(Z1)
+
+        # split the reconstructions up into binary and continuous components
+        Xr0_logits, Xr0_continuous = self.split_into_binary_and_continuous(Xr0) 
+        Xr1_logits, Xr1_continuous = self.split_into_binary_and_continuous(Xr1)
+
+        # do the same for the true data
+        X0_binary, X0_continuous = self.split_into_binary_and_continuous(self.X)
+        X1_binary, X1_continuous = self.split_into_binary_and_continuous(self.followup_X)
```
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: done.

File: general_autoencoder.py [hunk 8]
```diff
@@ -24,17 +24,22 @@ def __init__(self,
...
+        can_calculate_Z_mu=True,
+        need_ages=False,
         regularization_weighting_schedule={'schedule_type':'constant', 'constant':1}):
 
-        self.need_ages = False # whether ages are needed to compute loss or other quantities. 
+        self.need_ages = need_ages # whether ages are needed to compute loss or other quantities. 
         assert age_preprocessing_method in ['subtract_a_constant', 'divide_by_a_constant']
         self.age_preprocessing_method = age_preprocessing_method
-        self.include_age_in_encoder_input = include_age_in_encoder_input         
+        self.include_age_in_encoder_input = include_age_in_encoder_input
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: Ok, I will add this as an issue! I don't think I entirely understand how you want to change it.

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 9]
```diff
@@ -72,50 +71,59 @@ def _train_epoch(self, regularization_weighting):
...
+              "%i longitudinal batches of size %i") % (n_cross_sectional_batches,
+                                                       self.batch_size,
+                                                       n_longitudinal_batches, 
+                                                       self.longitudinal_batch_size))
+        
+        # generate a vector that indicates whether we're doing a longitudinal or 
+        # cross-sectional update at each step. 
+        batch_is_longitudinal = ([False for i in range(n_cross_sectional_batches)] + 
+                                 [True for i in range(n_longitudinal_batches)])
+        np.random.shuffle(batch_is_longitudinal)
```
[the remark this answers is not in the document]
Comment (epierson9) [reply variational_longitudinal_rate_of_aging_autoencoder.py]: I think we want a longitudinal batch size parameter because we may need it to control noise on the gradient update. If we're worried about washing out the longitudinal updates, we can just reduce the gradient batch size and/or increase the weighting? Either way, it seems like we want two parameters to control both noise + relative longitudinal weighting.
